lib/defender/aws_iot_defender_reporter_socket.c

```diff
@@ -0,0 +1,255 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/* Defender internal include. */
+#include "private/aws_iot_defender_internal.h"
+
+/* Specific metrics include. */
+#include "aws_iot_defender_metrics_socket.h"
+
+/* Secure sockets include. */
+#include "aws_secure_sockets.h"
+
+#define _TCP_CONN_TAG         AwsIotDefenderInternal_SelectTag( "tcp_connections", "tc" )
+#define _EST_CONN_TAG         AwsIotDefenderInternal_SelectTag( "established_connections", "ec" )
+#define _TOTAL_TAG            AwsIotDefenderInternal_SelectTag( "total", "t" )
+#define _CONN_TAG             AwsIotDefenderInternal_SelectTag( "connections", "cs" )
+#define _LOCAL_PORT_TAG       AwsIotDefenderInternal_SelectTag( "local_port", "lp" )
+#define _REMOTE_ADDR_TAG      AwsIotDefenderInternal_SelectTag( "remote_addr", "rad" )
+
+#define _LIS_TCP_PORTS_TAG    AwsIotDefenderInternal_SelectTag( "listening_tcp_ports", "tp" )
+#define _PORTS_TAG            AwsIotDefenderInternal_SelectTag( "ports", "pts" )
+#define _PORT_TAG             AwsIotDefenderInternal_SelectTag( "port", "pt" )
+
+/*-----------------------------------------------------------*/
+
+AwsIotSerializerError_t AwsIotDefenderInternal_GetTcpConnMetrics( AwsIotSerializerEncoderObject_t * pEncoderObject )
+{
+    AwsIotSerializerError_t error;
+
+    AwsIotSerializerEncoderObject_t tcpConnectionMap = AWS_IOT_SERIALIZER_ENCODER_CONTAINER_INITIALIZER_MAP;
+    AwsIotSerializerEncoderObject_t establishedMap = AWS_IOT_SERIALIZER_ENCODER_CONTAINER_INITIALIZER_MAP;
+    AwsIotSerializerEncoderObject_t connectionsArray = AWS_IOT_SERIALIZER_ENCODER_CONTAINER_INITIALIZER_ARRAY;
+
+    uint32_t tcpConnFlag = _AwsIotDefenderMetricsFlag[ AWS_IOT_DEFENDER_METRICS_TCP_CONNECTIONS ];
+
+    uint8_t hasEstablishedConnections = ( tcpConnFlag & AWS_IOT_DEFENDER_METRICS_TCP_CONNECTIONS_ESTABLISHED ) > 0;
+    uint8_t hasConnections = ( tcpConnFlag & AWS_IOT_DEFENDER_METRICS_TCP_CONNECTIONS_ESTABLISHED_CONNECTIONS ) > 0;
+    uint8_t hasTotal = ( tcpConnFlag & AWS_IOT_DEFENDER_METRICS_TCP_CONNECTIONS_ESTABLISHED_TOTAL ) > 0;
+    uint8_t hasLocalPort = ( tcpConnFlag & AWS_IOT_DEFENDER_METRICS_TCP_CONNECTIONS_ESTABLISHED_LOCAL_PORT ) > 0;
+    uint8_t hasRemoteAddr = ( tcpConnFlag & AWS_IOT_DEFENDER_METRICS_TCP_CONNECTIONS_ESTABLISHED_REMOTE_ADDR ) > 0;
+
+    uint8_t total = AwsIotDefenderInternal_GetEstablishedConnectionsCount();
+
+    _defenderMetricsConnection_t * pConnections = NULL;
+
+    /* 15 for IP + 1 for ":" + 5 for port + 1 terminator */
+    char remoteAddr[ 15 + 1 + 5 + 1 ] = "";
```

Comment:
I would declare a macro at the top of the file to track the value 15 + 1 + 5 + 1

Reply:
ok, I will do.